Add parallel H(curl) NURBS support for IGA electromagnetic problem and nurbs_ex25p

kaipengliuHIT opened the pull request:
**Note: This PR continues the work from #4441, which was closed due to branch force-push. Please refer to #4441 for previous discussion history.**

---

## Summary

This PR adds support for parallel H(curl) NURBS finite element spaces for isogeometric analysis (IGA) electromagnetic optimization applications, and add AMS Preconditioner Support for NURBS H(curl) Elements.

## Changes

- **Added VNURBSext support** in `ParFiniteElementSpace` to handle vector NURBS extensions for H(curl) spaces
- **Extended ParNURBSExtension constructor** to support vector NURBS spaces with proper parallel partitioning  
- **Added global element DOF table methods** in `NURBSExtension` for improved DOF management
- **Added nurbs_ex25p example** demonstrating parallel H(curl) NURBS usage

## Modified Files

7 files changed across 2 commits:

- [fem/fespace.cpp](file:///home/kaipeng/code/mfem/fem/fespace.cpp), [fem/fespace.hpp](file:///home/kaipeng/code/mfem/fem/fespace.hpp) - Core finite element space modifications
- [fem/pfespace.cpp](file:///home/kaipeng/code/mfem/fem/pfespace.cpp) - Parallel finite element space with VNURBSext support
- [mesh/nurbs.cpp](file:///home/kaipeng/code/mfem/mesh/nurbs.cpp), [mesh/nurbs.hpp](file:///home/kaipeng/code/mfem/mesh/nurbs.hpp) - NURBS extension enhancements  
- [miniapps/nurbs/makefile](file:///home/kaipeng/code/mfem/miniapps/nurbs/makefile) - Build configuration
- [miniapps/nurbs/nurbs_ex25p.cpp](file:///home/kaipeng/code/mfem/miniapps/nurbs/nurbs_ex25p.cpp) - New parallel example (752 lines)


## Related Issues

Continues #4441
<!--GHEX{"author":"kaipengliuHIT","assignment":"2025-12-16T20:44:53","editor":"tzanio","id":5143,"reviewers":["mlstowell","dylan-copeland","psocratis"],"merge":"2026-01-06T20:44:53","approval":"2025-12-30T20:44:53"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#5143](https://github.com/mfem/mfem/pull/5143) | @kaipengliuHIT | @tzanio | @mlstowell + @dylan-copeland + @psocratis | 12/16/25 | ⌛due 12/30/25 | ⌛due 1/6/26 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

Commit:
Add VNURBSext support for parallel H(curl) NURBS spaces

- Add GetGlobalElementDofTable methods to NURBSExtension base class
- Add new ParNURBSExtension constructor for VNURBSext support
- Update ParFiniteElementSpace to handle VNURBSext initialization
- Add nurbs_ex25p parallel example demonstrating H(curl) NURBS FEM

Commit:
Fix type compatibility issues in nurbs_ex25p

- Replace hardcoded 'double' with 'real_t' for MFEM compatibility
- Update complex types to use 'complex<real_t>' instead of 'complex<double>'
- Fix Array2D type mismatches
- Fix Vector constructor usage with real_t arrays

Commit:
Fix complex number arithmetic type compatibility

- Replace pow(dxs[i], 2) with pow2(dxs[i]) to avoid type promotion to double
- Cast literal 1.0 to real_t for complex arithmetic operations
- Ensures consistent real_t precision throughout complex calculations

Commit:
Update .gitignore for nurbs_ex25p outputs

Add patterns to ignore nurbs_ex25p executable and generated files:
- nurbs_ex25p (executable)
- nurbs_ex25p.* (mesh outputs like nurbs_ex25p.mesh000000)
- nurbs_ex25p-* (grid function outputs like nurbs_ex25p-sol_i.gf000000)
- nurbs_ex25p_ParaView/ (ParaView directory)

Commit:
Use standard ParaView directory name for nurbs_ex25p

- Change ParaView output path from 'nurbs_ex25p_ParaView' to 'ParaView'
- Follow MFEM convention for ParaView directory naming
- Update .gitignore to remove specific nurbs_ex25p_ParaView pattern
- Standard 'miniapps/nurbs/ParaView' pattern already covers all outputs

Commit:
Add nurbs_ex25p outputs to clean-exec target

- Add patterns to remove nurbs_ex25p.mesh* and nurbs_ex25p-* files
- Use nurbs_ex25p.mesh* instead of nurbs_ex25p.* to avoid removing .cpp
- Ensures make distclean properly removes all nurbs_ex25p outputs
- Required for gitignore test compliance

IdoAkkerman commented:
In PR #5098 I tried implementing the same feature. I followed a slightly different route to solve the issue.

kaipengliuHIT commented:
> In PR #5098 I tried implementing the same feature. I followed a slightly different route to solve the issue.

This PR is a resubmission of #4441  with a refactored code style. I believe my modifications are more concise and effective, offering greater scalability and compatibility in the future. They only introduce a minimal amount of code to the original base without altering any major logical structures. Moreover, it will remain compatible with parallelization on the GPU in the future.

Commit:
Add AMS preconditioner support for NURBS H(curl) elements

- Implement ProjectGrad and Project methods for NURBS_HCurl2D/3DFiniteElement
  * ProjectGrad: Projects H1 basis gradients onto edge tangent directions
  * Project: Builds H1-to-HCurl interpolation matrix for AMS V-cycle
- Add NURBS space detection in HypreAMS::MakeGradientAndInterpolation
  * Automatically creates NURBSFECollection H1 space for gradient operator
  * Handles NURBS control points via NURBSExtension framework
- Add AMS/AMG solver selection to nurbs_ex25p example
  * -ams flag (default): Use AMS preconditioner optimized for H(curl)
  * -amg flag: Use BoomerAMG for comparison

Performance on square-nurbs.mesh (order 2, 2 MPI ranks):
- AMS: 183 GMRES iterations (recommended)
- AMG: 287 GMRES iterations (+57% slower)

Depends-on: PR #5143 (VNURBSext support for parallel H(curl) NURBS spaces)

kaipengliuHIT committed:
Merge branch 'master' into mfem_for_IGA_EMopt

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

dylan-copeland commented on miniapps/nurbs/makefile:
```diff
@@ -22,7 +22,7 @@ MFEM_LIB_FILE = mfem_is_not_built
 SEQ_MINIAPPS = nurbs_ex1 nurbs_patch_ex1 nurbs_ex3 nurbs_ex5 nurbs_ex24 \
    nurbs_curveint nurbs_printfunc nurbs_solenoidal nurbs_naca_cmesh nurbs_surface
-PAR_MINIAPPS = nurbs_ex1p nurbs_ex11p
+PAR_MINIAPPS = nurbs_ex1p nurbs_ex11p nurbs_ex25p
```
In this directory, `CMakeLists.txt` should also be updated with `nurbex_ex25p`.

kaipengliuHIT replied:
It has been updated

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+//               mpirun -np 4 nurbs_ex25p -m ../../data/square-nurbs.mesh -o 2
+//               mpirun -np 4 nurbs_ex25p -m ../../data/cube-nurbs.mesh
+
+// Description:  This example code solves a simple electromagnetic wave parallel IGA(NURBS)
```
MFEM has in informal (not enforced by CI) maximum of 80 columns per line.

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+      cout << " Converged: " << gmres.GetConverged() << "\n";
+      if (use_ams)
+      {
+         cout << "\nAMS preconditioner successfully used for NURBS HCurl space!" << endl;
```
Long line.

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+real_t epsilon = 1.0;
+real_t omega;
+int dim;
+bool exact_known = false;
```
Unused.

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+   a.AddDomainIntegrator(new CurlCurlIntegrator(restr_muinv),NULL);
+   a.AddDomainIntegrator(new VectorFEMassIntegrator(restr_omeg),NULL);
+
+   int cdim = (dim == 2) ? 1 : dim;
```
`const int cdim`

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+   ConstantCoefficient absomeg(pow2(omega) * epsilon);
+   RestrictedCoefficient restr_absomeg(absomeg,attr);
+
+   ParBilinearForm prec(fespace);
```
It would be helpful to add some comments about the preconditioner assembled here.

dylan-copeland commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,99 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
    }
 }
 
+// ProjectGrad implementation for NURBS_HCurl2DFiniteElement
```
This comment is not informative.

kaipengliuHIT replied:
I removed this unnecessary comment.

dylan-copeland commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -1008,6 +1101,110 @@ void NURBS_HCurl3DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+      int sdim = Trans.GetSpaceDim();
+      real_t vk[Geometry::MaxDim];
+      Vector shape(fe.GetDof());
+      real_t tk_3d[9] = {1.0, 0.0, 0.0,  0.0, 1.0, 0.0,  0.0, 0.0, 1.0};  // Unit tangents
```
`constexpr real_t`

kaipengliuHIT replied:
I have modified the code as you suggested.

dylan-copeland replied:
Thanks! This line is long and can be split into 2 lines.

Commit:
Fix glvis parallel loading for NURBS H(curl)/H(div)

This commit addresses the segmentation fault when loading parallel
NURBS H(curl) output files in glvis, as reported in the PR feedback.

Changes:
1. Simplify output file naming in nurbs_ex25p.cpp to match ex25p.cpp
2. Initialize sFE/qFE/hFE pointers to nullptr in constructors
3. Add null pointer checks in NURBSExtension::LoadFE and LoadBE
4. Call SetDim() in GridFunction merge constructor for parallel loading

Root cause:
The GridFunction merge constructor (used during parallel loading) was
creating a new FiniteElementCollection but not calling SetDim(),
leaving sFE/qFE/hFE pointers uninitialized for NURBS_HCurlFECollection
and NURBS_HDivFECollection.

Testing:
- Serial loading: glvis -m mesh.000000 -g ex25p-sol_r.000000 ✓
- Parallel loading: glvis -np 2 -m mesh -g ex25p-sol_r ✓

kaipengliuHIT commented:
Thank you for the feedback! I've fixed the glvis parallel loading issue.

Changes made:
1. Simplified output file naming in nurbs_ex25p.cpp to match ex25p.cpp
2. Initialized sFE/qFE/hFE pointers to nullptr in NURBS_HCurlFECollection and NURBS_HDivFECollection constructors
3. Added null pointer checks in NURBSExtension::LoadFE() and LoadBE()
4. Fixed parallel loading by calling SetDim() in GridFunction merge constructor

Root cause:
The GridFunction merge constructor (used during parallel loading) was creating a new FiniteElementCollection but not calling SetDim(), leaving sFE/qFE/hFE pointers uninitialized for NURBS H(curl)/H(div) collections.

Testing:
Serial loading: `glvis -m mesh.000000 -g ex25p-sol_r.000000`
Parallel loading: `glvis -np 2 -m mesh -g ex25p-sol_r`

I'm now working on addressing the other feedback items. Will update soon.

Commit:
Update .gitignore for nurbs_ex25p output files

Add patterns for ex25p-sol_r.* and ex25p-sol_i.* to match the
simplified output file naming convention.

Commit:
Make GetGlobalElementDofTable methods virtual in NURBSExtension

Address reviewer feedback: the GetGlobalElementDofTable methods were
duplicated in both NURBSExtension (base class) and ParNURBSExtension
(derived class).

Changes:
- Make GetGlobalElementDofTable methods virtual in NURBSExtension base class
- Move these methods from private to protected in ParNURBSExtension
- Add override keyword to ParNURBSExtension methods

This allows proper polymorphic behavior when calling these methods through
NURBSExtension* pointers (e.g., in VNURBSExt arrays), while maintaining
the existing ParNURBSExtension implementation for parallel cases.

Commit:
Add test files and scripts to .gitignore

Prevent CI gitignore test failures by ignoring:
- Test executables (test_*)
- Shell scripts (*.sh)

These are local development files that should not be committed.

Commit:
Fix gitignore to include ex25p output files with numeric extensions

Commit:
Fix duplicate entries in .gitignore and ensure nurbs ex25p output files are properly ignored

Commit:
Fix comment in nurbs_ex25p.cpp

Commit:
Trigger CI check for gitignore fixes

Commit:
Update gitignore to use ex25p-*.* pattern for broader matching of nurbs_ex25p output files

kaipengliuHIT committed:
Merge branch 'master' into mfem_for_IGA_EMopt

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+//                                MFEM Example nurbs_25p -- modified for NURBS FE
```
I have revised.

Commit:
Temporarily comment out output file generation in nurbs_ex25p to test gitignore CI issue

Commit:
Complete merge with nurbs_ex25p output file comment

Commit:
Restore output file generation in nurbs_ex25p - files should be ignored by ex25p-*.* rule in .gitignore

Commit:
Update gitignore pattern for ex25p files to ex25p*.* for broader matching

Commit:
Refine gitignore pattern to specifically match nurbs_ex25p solution files: ex25p-sol*.*

Commit:
Change nurbs_ex25p output format to use fixed filenames instead of parallel numbered files to match other nurbs programs

dylan-copeland commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+   }
+
+   // 13. Save the refined mesh and the solution in parallel. This output can be
+   //     viewed later using GLVis: "glvis -np <np> -m mesh -g sol".
```
Are these changes committed? In the current version, the MPI rank is not in the mesh filename, so glvis only shows the subdomain for rank 0.

dylan-copeland replied:
Also, the glvis output should be done `if (visualization)`.

kaipengliuHIT replied:
Actually, the submitted code doesn't use the method described above to enable glvis. I will write a new reply later.

kaipengliuHIT replied:
It is now possible to display results in parallel using GLVis.

dylan-copeland replied:
Thank you for fixing the glvis output. A minor suggestion, for consistency with other MFEM examples, is to change `refined.mesh.` to `refined.`, `sol_r.gf.` to `sol_r.`, etc.

Commit:
Add unit tests for NURBS_HCurl ProjectGrad and Project functions

Add comprehensive unit tests for NURBS_HCurl2D and NURBS_HCurl3D
ProjectGrad and Project functions as requested for PR#5143.

Tests include:
- NURBS ProjectGrad 2D: Basic 2D NURBS_HCurl ProjectGrad test
- NURBS ProjectGrad 3D: Basic 3D NURBS_HCurl ProjectGrad test
- NURBS ProjectGrad Detailed 2D: More detailed verification for 2D
- NURBS ProjectGrad Detailed 3D: More detailed verification for 3D
- NURBS Project 2D: Test for 2D NURBS_HCurl Project function
- NURBS Project 3D: Test for 3D NURBS_HCurl Project function

Commit:
Apply code style formatting to test_nurbs_projectgrad.cpp

Commit:
Fix NURBS mesh requirement in tests

Update tests to use NURBS meshes by calling mesh.SetCurvature()
since NURBS finite element spaces require NURBS meshes.

Commit:
Fix NURBS mesh validation and path issues in tests

- Add NURBS extension validation check to skip tests if mesh is not actually a NURBS mesh
- Fix data file paths to use 'data/' instead of '../data/' for correct access from test directory
- Add null pointer checks for safer element access

Commit:
Add cube-nurbs.mesh and square-nurbs.mesh to unit test data

Commit:
Update mesh file paths in test_nurbs_projectgrad.cpp to use tests/unit/data/ directory

Commit:
Fix mesh file paths in test_nurbs_projectgrad.cpp to use data/ directory for CI compatibility

Commit:
Update fe_nurbs.cpp: Remove comments and make variables constexpr

Commit:
Fix fe_nurbs.cpp: Revert vk arrays to non-constexpr while keeping tk arrays as constexpr

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,99 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+      int sdim = Trans.GetSpaceDim();
+      real_t vk[Geometry::MaxDim];
+      Vector shape(fe.GetDof());
+      real_t tk_2d[4] = {1.0, 0.0, 0.0, 1.0};  // Unit tangents in x and y
```
I have modified the code as you suggested.

Commit:
Update nurbs.cpp and nurbs.hpp files

Commit:
Add nurbs_ex25p target to CMakeLists.txt

Commit:
Fix nurbs_ex25p build configuration and improve GLVis visualization

- Move nurbs_ex25p from non-MPI to MPI section in CMakeLists.txt
- Remove animation loop from GLVis visualization to prevent crashes
- Add proper MPI barriers for visualization synchronization
- Update output filenames with rank-specific suffixes

Commit:
Fix nurbs_ex25p code style and gitignore issues

- Fix code style issue: remove extra blank line that failed astyle check
- Update .gitignore to properly ignore numbered output files from nurbs_ex25p
- Fix refined mesh and solution file patterns to match numbered outputs
- Keep complete GLVis visualization functionality including animation

Commit:
Update .gitignore to include nurbs generated files and .clangd

Commit:
Fix clean-exec rule in nurbs makefile

- Replace incorrect nurbs_ex25p.mesh* nurbs_ex25p-* pattern
- Add proper patterns for refined.mesh.*, sol_r.gf.*, sol_i.gf.*
- This fixes cleanup of files generated by nurbs_ex25p execution

Commit:
Add missing file patterns to clean-exec rule

- Include sol_p.gf and sol_u.gf files generated by nurbs_ex5 and nurbs_solenoidal
- These files were previously not cleaned by the clean-exec target

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,804 @@
+
+
+   // 5. Define a finite element space on the mesh. Here we use the
+   //    Raviart-Thomas finite elements of the specified order.
```
Yes, it has been removed

Commit:
Add sol_u.gf and sol_p.gf to .gitignore

kaipengliuHIT committed:
Merge branch 'master' into mfem_for_IGA_EMopt

kaipengliuHIT committed:
Merge branch 'master' into mfem_for_IGA_EMopt

dylan-copeland commented on tests/unit/fem/test_nurbs_projectgrad.cpp:
```diff
@@ -0,0 +1,418 @@
+// Copyright (c) 2010-2025, Lawrence Livermore National Security, LLC. Produced
```
2026

kaipengliuHIT replied:
Thanks for the note. I updated the copyright header in the new test file accordingly, and the copyright CI now passes.

dylan-copeland commented on mesh/nurbs.hpp:
```diff
@@ -1027,6 +1027,13 @@ class NURBSExtension
+
+protected:
+   /// Construct and return a table of DOFs for each global element.
+   Table *GetGlobalElementDofTable();
```
Thanks for simplifying this. At this point, the `virtual` keyword can be removed, but a more important issue is whether these global tables can be avoided entirely. In 2D, we already have `Generate2DElementDofTable`, which sets the table for all active elements (all local elements in the parallel partition of the mesh). This PR introduces `Get2DGlobalElementDofTable`, which is essentially the same function but for all global elements. One concern is the scalability of using all global elements on each MPI rank. The global table is constructed by `GetGlobalElementDofTable` only in a `ParNURBSExtension` constructor. Are all the elements necessary, or would the active elements be sufficient, so that we could just call `GenerateElementDofTable`?

IdoAkkerman replied:
@dylan-copeland There is indeed some almost doubling up of code going on. Which at this point is unfortunately necessary. Ideally we have  `Generate2DElementDofTable` defined in `NURBSExtension` and one overloaded in `ParNURBSExtension`, this way the global Table could be obtained by calling the routine for the `NURBSExtension::Generate2DElementDofTable` this would require a rewrite of the NURBS classes (see #5105 )

IdoAkkerman commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,97 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+      int sdim = Trans.GetSpaceDim();
+      real_t vk[Geometry::MaxDim];
+      Vector shape(fe.GetDof());
+      constexpr real_t tk_2d[4] = {1.0, 0.0, 0.0, 1.0};  // Unit tangents in x and y
```
This looks like a convoluted way to define 2 normals. Spliting the dof loop in two explicit loops will allow you to avoid this.

kaipengliuHIT replied:
Addressed. I split the loop into two explicit component-wise loops, so the previous indirect normal/tangent selection logic is no longer needed. The first loop handles the x-directed block and uses the x reference direction directly. The second loop handles the y-directed block and uses the y reference direction directly.

IdoAkkerman commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,97 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+void NURBS_HCurl2DFiniteElement::Project(const FiniteElement &fe,
+                                         ElementTransformation &Trans,
+                                         DenseMatrix &I) const
+{
```
Please explain the assumptions/theory underlying this routine, ideally with a meaningful reference to a journal paper. I can imagine that this paper does not exist, in that case I think it would be worth writing it, however, personally I think the approach might have some issues...

Please see   PR #4326 for discussions  on "Nodal projections" of standard NURBS elements.
Likely there are analogies to be seen. In this PR we  ended settling on 3 projection options.
- Global L2 projection
- Local L2 projection -- based on the Bezier projection paper, with favorable and provable convergence properties
- Nodal _projection_ -- based on my intuition, which is formally not a projection, seems to have issues  H1 convergence, but results in monotone solutions

This routine seems closely related to the problematic NodalProjection.

kaipengliuHIT replied:
Thank you for the detailed explanation. I agree that this should not be presented as a global or local L2 projection with the same theoretical guarantees. The routine is intended as a local interpolation/projection operator consistent with the tensor-product NURBS H(curl) component spaces. The underlying H(curl) space structure follows the isogeometric de Rham construction described in Buffa, Sangalli, and Vázquez, "Isogeometric analysis in electromagnetics: B-splines approximation", CMAME 199 (2010), 1143-1152. In this implementation, the H(curl) basis is organized in component blocks, and the projector evaluates the corresponding scalar shape data at the component nodes and applies the H(curl) covariant mapping through the element Jacobian.

kaipengliuHIT replied:
I agree that this is closer in spirit to a nodal/interpolation-type projection than to a provably stable L2 projection. To reduce the risk of hidden issues, I added projector-level tests that do not go through the Krylov solver: exact-value tests for linear gradients and constant vector fields, plus a small 2D L2 refinement convergence test for `ProjectGrad`. So this PR now documents the assumptions and adds regression/convergence coverage, but I am not claiming a full theoretical proof for this projector. If a stronger projection with provable L2 stability is preferred, I think that would be a separate follow-up design, similar to the projection options discussed in PR #4326.

IdoAkkerman commented on mesh/nurbs.hpp:
```diff
@@ -1073,6 +1076,13 @@ class ParNURBSExtension : public NURBSExtension
+       The @a parent can be either a local NURBSExtension or a global one. */
+   ParNURBSExtension(NURBSExtension *parent,
+                     Array<NURBSExtension *> VNURBSExt,
+                     const ParNURBSExtension *par_parent);
```
In my opinion this constructor has a strange signature and strange behavior.
I can understand that it seems necessary but this is a consequence of how the `NURBSExtensions` are handled in `ParFiniteElementSpace::ParInit`,
 leading to unnecessary spaghetti code.

kaipengliuHIT replied:
Okay Prof Akkerman, I'll organize it later and create a new branch. Currently, it is possible to handle the multi-patch case with identical NURBS boundaries, meaning that in the mesh, each adjacent pair of NURBS patches share a common NURBS boundary.

Commit:
Address NURBS Hcurl review comments

Commit:
Merge upstream master into PR 5143

Commit:
Fix NURBS projector test copyright year

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,871 @@
+//
+// Compile with: make nurbs_ex25p
+//
+// Sample runs:  mpirun -np 4 nurbs_ex25p -m ../../data/square-nurbs.mesh
```
Yes, the large iteration count was caused by the overly large default refinement level. I changed the default `ref_levels` from 8 to 3 to keep the sample run reasonably small for regression testing, while still allowing finer runs through `-r/--refine`. The preconditioner work for H(curl) IGA electromagnetic problems has already been completed on my side. It is more suitable than the current AMS option for the IGA electromagnetic problems targeted here. After this PR is finished, I plan to submit it as a separate follow-up PR, since it is a separate solver/preconditioner topic from the NURBS H(curl) support added in this PR.

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,99 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
 }
 
+// ProjectGrad implementation for NURBS_HCurl2DFiniteElement
+void NURBS_HCurl2DFiniteElement::ProjectGrad(const FiniteElement &fe,
```
Addressed partially by adding a projector-level convergence test, independent of the Krylov solve.
The new `NURBSProjectGradConvergence2D` test computes the H(curl) field produced directly by `ProjectGrad` and compares it with the exact gradient of a smooth scalar function in the L2 norm. The test then checks monotone error reduction under uniform refinement.
This is intended to avoid relying only on the electromagnetic solve, where solver behavior could potentially hide projector issues. I agree that this is not a complete theoretical proof of the projector, but it provides a direct mathematical regression check on the projector itself.
The observed errors decrease under refinement:
  err0 = 2.2214414680
  err1 = 2.0882125885
  err2 = 1.1397362748
  err3 = 0.5659108383

Commit:
Initialize NURBS HCurl element nodes

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,97 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+   MFEM_ASSERT(fe.GetDim() == 2, "Dimension mismatch in ProjectGrad");
+
+   // Get the integration points (nodes) of this HCurl element
+   const IntegrationRule &ir = Nodes;
```
Addressed. The H(curl) NURBS `Nodes` are now explicitly initialized, so they are not left as the default trivial points. The nodes are set as tensor-product reference points following the same component-block ordering used by the H(curl) basis. In 2D, the first block corresponds to the x-directed component space and the second block to the y-directed component space. In 3D, the same structure is used for the x-, y-, and z-directed blocks. I also added unit-test checks to verify that the NURBS H(curl) nodes have the expected size, lie in the reference element, and are not all zero.

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,97 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+   grad.SetSize(dof, fe.GetDof());
+
+   // For each DOF in the HCurl space
+   for (int k = 0; k < dof; k++)
```
Addressed. I rewrote this routine using two explicit component-wise loops. The first loop handles the x-directed H(curl) block with orders `(p_x, p_y+1)`, and the second loop handles the y-directed block with orders `(p_x+1, p_y)`. This removes the previous per-dof branch logic and makes the routine consistent with the other `NURBS_HCurl2DFiniteElement` tensor-product loops.

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,97 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+   // For each DOF in the HCurl space
+   for (int k = 0; k < dof; k++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(k);
```
Addressed. I added explicit comments explaining the component-block ordering and initialized the H(curl) NURBS `Nodes` consistently with that ordering. The ordering is intentional: the NURBS H(curl) basis is stored by component blocks. In 2D, the first block is x-directed with orders `(p_x, p_y+1)`, and the second block is y-directed with orders `(p_x+1, p_y)`. The `Nodes` now follow the same block structure, so the first block is used for the dx/x-directed component and the second block for the dy/y-directed component. I also added unit-test checks to verify that these nodes are explicitly initialized and are not left as trivial default points.

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -817,6 +817,97 @@ void NURBS_HCurl2DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+      constexpr real_t tk_2d[4] = {1.0, 0.0, 0.0, 1.0};  // Unit tangents in x and y
+
+      I.SetSize(dof, sdim*fe.GetDof());
+      for (int k = 0; k < dof; k++)
```
Addressed. I split this into explicit component-wise loops.

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -1008,6 +1099,108 @@ void NURBS_HCurl3DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+   grad.SetSize(dof, fe.GetDof());
+
+   // For each DOF in the HCurl space
+   for (int k = 0; k < dof; k++)
```
Addressed. I split the 3D routine into three explicit component-wise loops.

kaipengliuHIT commented on fem/fe/fe_nurbs.cpp:
```diff
@@ -1008,6 +1099,108 @@ void NURBS_HCurl3DFiniteElement::CalcCurlShape(const IntegrationPoint &ip,
+      constexpr real_t tk_3d[9] = {1.0, 0.0, 0.0,  0.0, 1.0, 0.0,  0.0, 0.0, 1.0};  // Unit tangents
+
+      I.SetSize(dof, sdim*fe.GetDof());
+      for (int k = 0; k < dof; k++)
```
Addressed.

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,871 @@
+   Array<int> ess_bdr(pmesh->bdr_attributes.Max());
+   ess_bdr = 1;
+   fespace->GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
+   cout << "Number of knowns in essential BCs: "
```
Addressed. The essential-BC output now includes the MPI rank, since that value is rank-local. The linear-system size output is now restricted to rank 0, following the usual example output pattern.

kaipengliuHIT commented on miniapps/nurbs/nurbs_ex25p.cpp:
```diff
@@ -0,0 +1,871 @@
+   // Enable hardware devices such as GPUs, and programming models such as
+   // CUDA, OCCA, RAJA and OpenMP based on command line options.
+   Device device(device_config);
+   device.Print();
```
Addressed. `device.Print()` is now called only on rank 0.

kaipengliuHIT commented on mesh/nurbs.cpp:
```diff
@@ -5638,146 +5790,187 @@ ParNURBSExtension::ParNURBSExtension(NURBSExtension *parent,
+
+   MFEM_VERIFY(par_parent->partitioning,
+               "parent ParNURBSExtension has no partitioning!");
+
```
I agree with this concern. The current constructor is doing orchestration for the H(curl) component extensions, and that responsibility would fit better in `ParFiniteElementSpace::ParInit` or in a dedicated helper used from there. For this PR, I added comments and guards clarifying that the current path is limited to the single-patch H(curl) case and should not be treated as the final multipatch design. A cleaner design would move the component-extension orchestration and group/table construction out of `ParNURBSExtension`, preferably avoiding global element tables on each rank. I would prefer to keep that refactor as follow-up work, because it touches the parallel NURBS initialization design more broadly than the single-patch H(curl) support added in this PR.

kaipengliuHIT commented:
> > I believe my modifications are more concise and effective, offering greater scalability and compatibility in the future.
> 
> Can you corroborate?
> 
> I think the handling of the `NURBSExtensions` in `ParFiniteElementSpace::ParInit`, and the resulting `ParNURBSExtension` will paint us in a corner when trying to extend to the multipatch case.

Thank you for pointing this out. I agree that the current handling of the component `NURBSExtension`s in `ParFiniteElementSpace::ParInit` and the resulting `ParNURBSExtension` is not the final design I would want for generic multipatch H(curl) support. 
For this PR, I have clarified the scope in the code: the current H(curl) extension path is guarded to single-patch NURBS meshes, and the component-wise table construction is documented as a single-patch mechanism, not as a complete multipatch continuity/orientation solution.
I do have follow-up multipatch H(curl) work locally, but I agree that the generic multipatch case needs a cleaner design for tangential DOF matching, orientation, and patch-interface handling. I would prefer to keep that as a separate follow-up PR, so this PR remains focused on parallel single-patch H(curl) NURBS support and the nurbs_ex25p miniapp.

kaipengliuHIT commented:
Thank you all for the detailed review. I have pushed another update addressing the main implementation and testing comments.

Summary of the latest changes:
  - Rewrote the NURBS H(curl) `ProjectGrad` and matrix-form `Project` routines using explicit component-wise loops in 2D and 3D.
  - Added comments explaining the tensor-product H(curl) component-block ordering.
  - Explicitly initialized the NURBS H(curl) `Nodes` so they are not left as trivial default points.
  - Strengthened the unit tests so they check projected values, not only successful execution.
  - Added a small 2D `ProjectGrad` L2 refinement convergence test.
  - Cleaned up the `nurbs_ex25p` miniapp output/default refinement behavior.
  - Merged current `mfem/master` with a regular merge commit, without force-pushing.

  I also want to clarify the scope of this PR. The current implementation is intended for the single-patch parallel H(curl) NURBS case. I agree with the review comments that a generic
  multipatch H(curl) implementation needs a cleaner design for tangential DOF matching, orientation, patch-interface handling, and possibly avoiding global element-dof tables on each rank.
  I have therefore added comments/guards to make the current limitation explicit rather than presenting this as the final multipatch architecture.

  After this PR is finished, I plan to submit two follow-up PRs:
  1. A multipatch H(curl) NURBS version.
  2. A solver/preconditioner update for H(curl) IGA electromagnetic problems. That preconditioner work has already been completed on my side and is more suitable than the current AMS option for the IGA electromagnetic problems targeted here.

I think keeping those as follow-up PRs will make this review more focused and avoid mixing the single-patch H(curl) support, generic multipatch design, and solver/preconditioner changes in one PR.

kaipengliuHIT commented:
For additional context, the follow-up solver/preconditioner work is currently in my `opt_em_iga` project:
  https://github.com/kaipengliuHIT/opt_em_iga.
  The multipatch H(curl) NURBS work is currently on the `multipatch_hcurl_fix_backup_20260523` branch. After
  this PR is finished, I plan to clean up these two pieces of work and submit them as separate follow-up
  PRs.
